### PR TITLE
feat: add batchAccept route to accept multiple hackers at once

### PR DESCRIPTION
**Pull request by ManethKulatunge**
### Tickets:

-   HCK-153

### List of changes:

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

-

## Type of change

Please delete options that are not relevant.
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

### How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Added many integration tests, and also did some manual postman tests.

### Checklist:

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [ ] Listed change(s) in the Changelog
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes
-   [ ] I have made corresponding changes to the documentation
-   [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by pierreTklein** [comment]
@loreina @logan-r please have a look now at the code. I finished this code, added tests, and also manually tested.

After this goes in, we will need to do the following:
`npm run seed` to add the new route
`npm run docs` to generate the docs

**Comment by krubenok** [comment]
🚨DANGER🚨

We still don't know how to repeatedly fix DB incompatibilities that we suspect are caused by 'npm run seed'. See the great chaos of production roll out circa December 2019.

**Comment by pierreTklein** [comment]
Lemme see if I can rewrite `npm run seed` to be less dangerous, lol

File: controllers/hacker.controller.js [hunk 1]
```diff
@@ -92,8 +92,19 @@ function sentDayOfEmail(req, res) {
...
+        data: {
+            success_ids: req.body.ids,
+            errors: req.errors || []
+        }
+    });
+}
+
 module.exports = {
     updatedHacker: updatedHacker,
+    completedBatchUpdate: completedBatchUpdate,
```
Comment (loreina) [inline controllers/hacker.controller.js:107]: nit: don't think this fits with how the other functions are named here. maybe `batchUpdatedHackers` but that sounds like a mouthful idk

File: routes/api/hacker.js [hunk 2]
```diff
@@ -326,19 +326,52 @@ module.exports = {
...
+                Middleware.Hacker.obtainEmailByHackerId,
+                Middleware.Hacker.completeStatusUpdateEmail,
+                Controllers.Hacker.updatedHacker
+            );
 
+        /**
+         * @api {patch} /hacker/batchAccept/ accept array of Hackers
+         * @apiName acceptHacker
+         * @apiGroup Hacker
+         * @apiVersion 2.0.0
```
Comment (loreina) [inline routes/api/hacker.js:346]: this will go into 3.0.0

File: package.json [hunk 3]
```diff
@@ -36,6 +36,7 @@
         "multer": "^1.4.2",
         "passport": "^0.4.0",
         "passport-local": "^1.0.0",
+        "promise.allsettled": "^1.0.2",
```
Comment (loreina) [inline package.json:39]: where's this used?

File: services/account.service.js [hunk 4]
```diff
@@ -20,6 +21,30 @@ function findById(id) {
...
+    const TAG = `[Account Service # findByHackerId]:`;
+    const query = {
+        _id: id
+    };
+    const hacker = await Hacker.findById(
+        query,
+        logger.queryCallbackFactory(TAG, "account", query)
+    ).populate({
+        path: "accountId",
+        select: " -password"
```
Comment (loreina) [inline services/account.service.js:40]: what does this do?

**Comment by loreina** [comment]
for `npm run docs` is there a way to trigger between v1, v2, v3? we'd probably have to do some back tracing to config that tho

**Comment by krubenok** [comment]
> we should only run it once a year right before apps launch.

I disagree. If we can gain confidence in `npm run seed` it should be part of the CI/CD for deploying to production.

**Comment by loreina** [comment]
accident 🙃

File: package.json (hunk 3 above)
Comment (pierreTklein) [reply package.json:39]: This is used in  `hacker.middleware.js` for `Promise.allSettled(...)`.

File: services/account.service.js (hunk 4 above)
Comment (pierreTklein) [reply services/account.service.js:40]: Selects all fields except for the password. We never want to populate the password unless we're doing a login / auth.

File: middlewares/hacker.middleware.js [hunk 5]
```diff
@@ -649,7 +785,7 @@ async function createHacker(req, res, next) {
         });
     }
     const hacker = await Services.Hacker.createHacker(hackerDetails);
-    if (!!hacker) {
```
Comment (loreina) [inline middlewares/hacker.middleware.js:652]: lol we really did this huh

Comment (pierreTklein) [reply middlewares/hacker.middleware.js:652]: Yep :/

**Comment by pierreTklein** [comment]
#652 fixes all of our worries about `npm run seed`. We should be able to integrate it with CI/CD in the future. 

That being said, there is a *brief* moment when running the script where NOBODY has permissions (we basically do a DROP TABLE; CREATE TABLE), so it's probably worth disabling `npm run seed` when apps are open / near deadline.

**Pull request merged**